lib/basedriver/capabilities.js

```
@@ -0,0 +1,92 @@
+                                              {fullMessages: false});
+
+  if (validationErrors) {
+    for (let [attribute, reasons] of _.toPairs(validationErrors)) {
+      for (let reason of reasons) {
+        message += ` ${attribute} ${reason},`;
+      }
+    }
+    message = `${message.slice(0, -1)}.`;
+    throw new Error(message);
+  }
+
+  // Return caps
+  return caps;
+}
+
+// Parse capabilities (based on https://www.w3.org/TR/webdriver/#processing-capabilities)
+function parseCapabilities (capabilities, alwaysMatch, firstMatch = [], constraints = {}) {
+  let capabilitiesRequest, requiredCapabilities, allFirstMatchCapabilities, validatedFirstMatchCapabilities, matchedCapabilities = null;
+  
+  capabilitiesRequest = capabilities;
+  if (!_.isObject(capabilitiesRequest)) {
+    throw new Error('The capabilities argument was not valid for the following reason(s): "capabilities" must be a JSON object.');
+  }
+
+  requiredCapabilities = alwaysMatch;
+  if (!_.isObject(requiredCapabilities)) {
+    requiredCapabilities = {}; 
+  }
+
+  allFirstMatchCapabilities = firstMatch;
+  if (!_.isArray(allFirstMatchCapabilities)) {
+    throw new Error('The firstMatch argument was not valid for the following reason(s): "firstMatch" must be a JSON array or undefined');
+  }
+
+  validatedFirstMatchCapabilities = allFirstMatchCapabilities.map((firstMatchCapabilities, index) => {
+    try {
+      return validateCapabilities(firstMatchCapabilities, constraints);
+    } catch (e) {
+      throw new Error(`The firstMatch argument was not valid for the following reason: firstMatch[${index}] ${e.message}`);
+    }
+  });
+
+  for (let firstMatchCapabilities of validatedFirstMatchCapabilities) {
+    try {
+      let mergedCapabilities = mergeCapabilities(requiredCapabilities, firstMatchCapabilities);
+      matchedCapabilities = matchCapabilities(mergedCapabilities);
+      if (matchedCapabilities) {
+        break;
+      }
+    } catch (ign) { }
```

why do we want to ignore an error in this part of the algorithm?